[DNF5] Replace multiple transaction actions by one

j-mracek opened the pull request:
It also allow to remove one attribute of TransactionPackage

dmach commented on include/libdnf/transaction/transaction_item_action.hpp:
```diff
@@ -28,19 +28,15 @@ namespace libdnf::transaction {
-    DOWNGRADED = 3,     // an original package version that was replaced
-    // TODO(jmracek) OBSOLETE action is not used in DNF-4 and DNF-5. What about to remove it?
-    OBSOLETE = 4,       //
-    OBSOLETED = 5,      //
     UPGRADE = 6,        //
-    UPGRADED = 7,       //
     REMOVE = 8,         // a package that was removed from the system
     REINSTALL = 9,      // a package that was reinstalled with the identical version
-    REINSTALLED = 10,   // a package that was reinstalled with the identical version (old repo, for example)
-    REASON_CHANGE = 11  // a package was kept on the system but it's reason has changed
+    REASON_CHANGE = 11, // a package was kept on the system but it's reason has changed
+    REPLACED = 12       // Replace actions UPGRADED, DOWNGRADED, REINSTALLED, OBSOLETED
 };
```
We'll probably need (just a detail, to avoid possible gaps in the numbering):
REASON_CHANGE_FROM=12
REPLACED=13

I'm not sure about removing the items from the enum; We'll probably figure out this when dealing with database migrations and other stuff.

j-mracek replied:
I've got a suggestion to include REASON_CHANGE_FROM when we will need it. Therefore I remove it from PR.

j-mracek committed:
Replace reverse transaction actions by REPLACED

It requires a change of the structure of transaction package where
the reverses package (e.g. DOWNGRADED) and obsoletes are stored in
replaces.

j-mracek committed:
Color differently TSI when obsoleted

It will color in brown name or arch of replaced package when replacing
package has a different name or arch.

It will provide an easy way how to find obsoletes in transaction table.

j-mracek committed:
Report replaced packages as replaced

lukash commented on dnfdaemon-client/wrappers/dbus_transaction_package_wrapper.hpp:
```diff
@@ -43,6 +43,8 @@ class DbusTransactionPackageWrapper {
     DbusPackageWrapper get_package() const { return package; }
     libdnf::transaction::TransactionItemAction get_action() const { return action; }
     libdnf::transaction::TransactionItemReason get_reason() const { return reason; }
+    // TODO(jmracek) Fully integrate get_replaces()
```
I don't mind this as a note for yourself if you're going to do this in near future, but otherwise a better description would be nice, because it's not clear at all what you mean by this.

j-mracek replied:
Fixed

lukash commented on include/libdnf/transaction/transaction_item_action.hpp:
```diff
@@ -28,19 +28,21 @@ namespace libdnf::transaction {
 // because removed items (RPMs) must be excluded from reason resolution:
 // * Package.cpp - Package::resolveTransactionItemReason
+// TODO(jmracek) Copy enum to SWDB including their string definitions and already removed values to keep compatibility
+// Then transform old database data to transform stored reverse action to poplaced action and remove all unused action from code
```
Typo: poplaced -> REPLACED (caps make it clearer)

j-mracek replied:
Fixed

lukash commented on libdnf/base/transaction.cpp:
```diff
@@ -495,22 +490,17 @@ void Transaction::Impl::set_transaction(rpm::solv::GoalPrivate & solved_goal, Go
-        for (int i = 1; i < obs.size(); ++i) {
-            rpm::Package obsoleted(base, rpm::PackageId(obs[i]));
+        for (auto replaced_id : obs) {
+            rpm::Package obsoleted(base, rpm::PackageId(replaced_id));
```
This code is awfully similar between the various actions, though there are slight differences, like the variable above is called `obsoleted` here and `replaced_pkg` in the downgrades block above. I'm not entirely sure but it appears this should be factored out into a function to deduplicate and unify the code.

j-mracek replied:
Correct, Upgrade, reinstall, and downgrade have the same code. I will think about it.

j-mracek replied:
Fixed

lukash commented on libdnf/transaction/transaction_item_action.cpp:
```diff
@@ -80,6 +82,9 @@ std::string TransactionItemAction_get_short(TransactionItemAction action) {
             return "?";
+        case TransactionItemAction::REPLACED:
+            // TODO(jmracek): replace "?" with something better
+            return "?";
```
While this will likely need revisiting, what about at least using a unique letter, e.g. "X", just in case it ever matters before we finalize these?

j-mracek replied:
What about to use "O" instead "X" or "?"?

lukash replied:
Sure. I assume we'll revisit these and figure out the best letters.

j-mracek replied:
Fixed

lukash commented on libdnf-cli/output/transaction_table.hpp:
```diff
@@ -294,6 +304,26 @@ bool print_transaction_table(Transaction & transaction) {
+            scols_cell_set_color(scols_line_get_cell(ln_replaced, COL_NAME), replaced_color);
+            scols_cell_set_color(scols_line_get_cell(ln_replaced, COL_REPO), replaced_color);
+            scols_cell_set_color(scols_line_get_cell(ln_replaced, COL_SIZE), replaced_color);
+        }
```
I'm not insisting on this, but it would be nice to have a scenario or two in the CI stack that would test this. It would already make my life easier as I wouldn't have to test this manually, I could just run the tests...

j-mracek committed:
Replace redundant code by method

It will make future changes easier.

lukash commented on libdnf/base/transaction.cpp:
```diff
@@ -372,6 +372,34 @@ std::string Transaction::all_package_solver_problems_to_string() {
+        Id id = list_for_tsi[index];
+        auto obs = solved_goal.list_obsoleted_by_package(id);
+        if (obs.empty()) {
+            throw LogicError(no_obsoletes_msg);
```
Are you sure this is a `LogicError`? I don't like it here. I understand if `obs` is empty it's an inconsistency in libsolv's output, but this is not something e.g. a user of the API can influence, it's just a libsolv bug. There is a huge amount of inconsistencies in libsolv's output which could occur, by which rule do we check some of them and not the others?

Do we really want to crash on this condition? I'd lean towards not checking this at all here, what do you think?

j-mracek replied:
Yes, I am sure that it is a logic Error. When it appears, it means something terribly wrong in libdnf or libsolv

j-mracek replied:
In dnf4 fore it ends in index exception in Python - 
`upgraded = obs.pop(0)`
`IndexError: pop from empty list`

lukash commented on libdnf/base/transaction.cpp:
```diff
@@ -372,6 +372,34 @@ std::string Transaction::all_package_solver_problems_to_string() {
     return output;
 }
 
+void Transaction::Impl::set_transaction_internal(
```
This method name is very bad. I have an idea how to refactor this a bit differently though, but instead of trying to explain, I'll give it a try myself and push a fixup commit, after we decide what to do about the `LogicError`.

lukash commented:
@j-mracek I've pushed a fixup for your last commit which I think shares the code for creating the transaction packages in a much better way. It should be possible to share this with the install packages too, but since you don't want to do it now, I've added a TODO. Please take a look and if you like it, squash it with your last commit. Thanks.

j-mracek commented:
> @j-mracek I've pushed a fixup for your last commit which I think shares the code for creating the transaction packages in a much better way. It should be possible to share this with the install packages too, but since you don't want to do it now, I've added a TODO. Please take a look and if you like it, squash it with your last commit. Thanks.

Thanks, I squashed your commit.
